libvast/src/segment.cpp

```diff
@@ -13,112 +13,117 @@
 
 #include "vast/segment.hpp"
 
-#include <caf/binary_deserializer.hpp>
-
 #include "vast/bitmap.hpp"
 #include "vast/bitmap_algorithms.hpp"
 #include "vast/detail/assert.hpp"
 #include "vast/detail/byte_swap.hpp"
 #include "vast/detail/narrow.hpp"
+#include "vast/error.hpp"
+#include "vast/fbs/segment.hpp"
+#include "vast/fbs/utils.hpp"
+#include "vast/filesystem.hpp"
 #include "vast/ids.hpp"
 #include "vast/logger.hpp"
 #include "vast/si_literals.hpp"
 #include "vast/table_slice.hpp"
 
+#include <caf/binary_deserializer.hpp>
+
 namespace vast {
 
 using namespace binary_byte_literals;
 
 segment_ptr segment::make(chunk_ptr chunk) {
   VAST_ASSERT(chunk != nullptr);
-  // Setup a CAF deserializer
-  caf::binary_deserializer source{nullptr, chunk->data(), chunk->size()};
-  auto result = segment_ptr{new segment, false};
-  if (auto error = source(result->header_, result->meta_)) {
-    VAST_ERROR_ANON(__func__, "failed to deserialize segment meta data");
+  // Verify flatbuffer integrity.
+  auto data = reinterpret_cast<const uint8_t*>(chunk->data());
+  auto verifier = flatbuffers::Verifier{data, chunk->size()};
+  if (!fbs::VerifySegmentBuffer(verifier)) {
+    VAST_ERROR_ANON(__func__, "failed to verify segment buffer");
     return nullptr;
   }
-  if (result->header_.magic != magic) {
-    VAST_ERROR_ANON(__func__, "got invalid segment magic",
-                    result->header_.magic, "instead of", magic);
+  // Perform version check.
+  auto ptr = fbs::GetSegment(chunk->data());
+  if (ptr->version() != fbs::SegmentVersion::v1) {
+    VAST_ERROR_ANON(__func__, "unsupported segment version");
     return nullptr;
   }
-  if (result->header_.version > version) {
-    VAST_ERROR_ANON(__func__, "got newer segment version",
-                    result->header_.version, "instead of", version);
-    return nullptr;
+  return segment_ptr{new segment{std::move(chunk)}, false};
+}
+
+uuid segment::id() const {
+  auto ptr = fbs::GetSegment(chunk_->data());
+  auto data = span<const uint8_t, 16>(ptr->uuid()->Data(), 16);
+  return uuid{as_bytes(data)};
+}
+vast::ids segment::ids() const {
+  vast::ids result;
+  auto ptr = fbs::GetSegment(chunk_->data());
```

Comment:
Could we make this a `mutable` member variable, or is it possible for the chunk to grow in code outside of this file and the `builder`?

Reply:
We could. I decided to avoid it because having just a chunk as member is kinda like pimpl.